### PR TITLE
Added JIT CNV4 compiler for ppc64le, ppc64 and ppc

### PR DESCRIPTION
**Pull request by nioroso-x3**
Initial version of JIT compiler for PPC, ported from my fork of XMRIG. Still requires testing.

### CODE REVIEW

File: src/crypto/CryptonightR_JIT.c [hunk 1]
```diff
@@ -117,7 +223,82 @@ int v4_generate_JIT_code(const struct V4_Instruction* code, v4_random_math_JIT_f
...
+    #if __BYTE_ORDER == __BIG_ENDIAN
+    //use end of buffer to store pointer to beginning of buffer (weirdness of the BE ppc abi)
+    size_t l = buf_size/sizeof(void*);
+    if (sizeof(void*) == sizeof(uint32_t)){
+        *(JIT_code+l-1) = JIT_code;
+    }
+    else if(sizeof(void*) == sizeof(uint64_t)){
+        uint64_t* fP = (uint64_t*)JIT_code;
+        *(fP+l-1) = (uint64_t)JIT_code;
+    }
```
Comment (moneromooo-monero) [inline src/crypto/CryptonightR_JIT.c]: That can also be a single op.

**Comment by nioroso-x3** [comment]
I'll fix this your comments and some extra bugs in a new pull, just reject this one.

**Comment by moneromooo-monero** [comment]
You should push if instead once you've fixed everything (and squash commits).

**Comment by nioroso-x3** [comment]
Test project /monero-m/build/Linux/master/debug
      Start  1: hash-target
 1/19 Test  #1: hash-target ......................   Passed    2.21 sec
      Start  2: core_tests
 2/19 Test  #2: core_tests .......................***Exception: Child terminated  5.26 sec //killed by me
      Start  3: cncrypto
 3/19 Test  #3: cncrypto .........................***Exception: Child terminated  7.59 sec //killed by me
      Start  4: cnv4-jit
 4/19 Test  #4: cnv4-jit .........................   Passed  751.12 sec
      Start  5: unit_tests
 5/19 Test  #5: unit_tests .......................***Exception: Child terminated212.90 sec //killed by me
      Start  6: difficulty
 6/19 Test  #6: difficulty .......................   Passed    0.08 sec
      Start  7: wide_difficulty
 7/19 Test  #7: wide_difficulty ..................   Passed   24.66 sec
      Start  8: block_weight
 8/19 Test  #8: block_weight .....................   Passed  108.16 sec
      Start  9: hash-fast
 9/19 Test  #9: hash-fast ........................   Passed    0.04 sec
      Start 10: hash-slow
10/19 Test #10: hash-slow ........................   Passed    0.63 sec
      Start 11: hash-slow-1
11/19 Test #11: hash-slow-1 ......................   Passed    0.69 sec
      Start 12: hash-slow-2
12/19 Test #12: hash-slow-2 ......................   Passed    1.75 sec
      Start 13: hash-slow-4
13/19 Test #13: hash-slow-4 ......................   Passed    1.78 sec
      Start 14: hash-tree
14/19 Test #14: hash-tree ........................   Passed    0.01 sec
      Start 15: hash-extra-blake
15/19 Test #15: hash-extra-blake .................   Passed    0.02 sec
      Start 16: hash-extra-groestl
16/19 Test #16: hash-extra-groestl ...............   Passed    0.02 sec
      Start 17: hash-extra-jh
17/19 Test #17: hash-extra-jh ....................   Passed    0.02 sec
      Start 18: hash-extra-skein
18/19 Test #18: hash-extra-skein .................   Passed    0.02 sec
      Start 19: hash-variant2-int-sqrt
19/19 Test #19: hash-variant2-int-sqrt ...........   Passed  462.76 sec

It's passing cnv4jit test and the hash_slow4 test now, and x86 is unaffected. I guess its ready for merging.

**Comment by moneromooo-monero** [comment]
How long did the jit test take before that ?

**Comment by nioroso-x3** [comment]
Double the time. I haven’t implemented altivec yet. For reference my xmrig fork went from 2 hash/s/core to 60 hash/s/core after optimizations.

**Comment by moneromooo-monero** [comment]
Can you squash those commits please ?

**Comment by nioroso-x3** [comment]
Hmm, how do I that? I just tried and it added two more commits.

**Comment by nioroso-x3** [comment]
Meh, I'll just delete the repo and refork, dunno what I did wrong.

**Comment by moneromooo-monero** [comment]
+duplicate

of https://github.com/monero-project/monero/pull/5543

**Pull request closed without merging**